refactor: internal ros communication

boczekbartek opened the pull request:
## Purpose

- Refactors internal ros communication and they way ros2-based tools communicate with the agent.

## Proposed Changes

- replace nodes spinning by the executor/rclpy by callbacks - from my testing this is the only reliable method for waiting on `rclpy.Future` for nodes that are being span by `executor.spin()`
- this allows to remove the dedicated node for tools that needed to spin on futures
- refactor `RaiNode` and extract methods to separate classes to improve code clarity
- improve feedback from `ros2 action`s by parsing `nav2` error codes

## Issues

- https://github.com/RobotecAI/rai/issues/313
   - in this PR an alternative approach described in this issue is used and extended

## Testing

- on rosbot-xl demo


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- **New Features**
	- Added support for transform checking in robot interactions
	- Introduced new tools for robot interfaces and transform retrieval

- **Improvements**
	- Enhanced logging for tool execution
	- Improved ROS 2 action and topic handling
	- Updated node discovery and asynchronous future result management

- **Technical Updates**
	- Refactored node and tool classes for better organization
	- Updated callback group handling
	- Modified error handling and logging mechanisms

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

boczekbartek committed:
fix(`GetTransform`): add missing `args_schema`

boczekbartek committed:
refactor(`RaiNode`): executors and subscribers

boczekbartek committed:
improve logging of tool calls

boczekbartek committed:
remove RaiAsctionToolsNode

boczekbartek committed:
refactor: executors and actions calling

boczekbartek committed:
further refactoring of `RaiNode`

boczekbartek committed:
revert(`config.toml`):

boczekbartek committed:
revert: executor to standard `rclpy.MultiThreadedExecutor`

- with current method based on callbacks the default executor is good

boczekbartek committed:
chore: remove unnecessary log

boczekbartek committed:
fix: import

boczekbartek committed:
feat: add function to get rclpy.Future result using callback

boczekbartek committed:
chore: add licence note

maciejmajek commented on src/rai/rai/utils/ros.py:
```diff
@@ -12,19 +12,60 @@
+        self.allowlist = allowlist
 
-    def set(self, topics, services, actions):
+        self.topcies_and_types: Dict[str, str] = dict()
```
```suggestion
        self.topics_and_types: Dict[str, str] = dict()
```

maciejmajek commented on examples/rosbot-xl-demo.py:
```diff
@@ -18,23 +18,22 @@
 import rclpy
 import rclpy.executors
 import rclpy.logging
-from rai_open_set_vision.tools import GetDetectionTool, GetDistanceToObjectsTool
```
Why?

boczekbartek replied:
I removed them, because these tools doesn't work any more (always say that distance equals to 0 meters). Didn't have time to debug it yet, but maybe they should be replaced with ones used in manipulation demo.

boczekbartek replied:
@maciejmajek I reverted this change and fixed the tools. `conversion_ratio` was wrong

maciejmajek reviewed (commented):
Tested on rosbot xl demo. Works as intended. Please respond to comments

boczekbartek committed:
Update src/rai/rai/utils/ros.py

Co-authored-by: Maciej Majek <46171033+maciejmajek@users.noreply.github.com>

boczekbartek committed:
fix(`rosbot_xl_demo`): revert detection removal and fix it

- rollback open-set detection removal
- fix it by setting conversion_ratio=1.0

boczekbartek commented:
@maciejmajek I responded to your comments. Could you check again?

boczekbartek commented:
@coderabbitai full review

boczekbartek committed:
improve ros discovery

boczekbartek committed:
Apply suggestions from code review

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

boczekbartek committed:
fix(`GetTransformTool`): make default values consistent

boczekbartek commented on src/rai/rai/tools/ros/native_actions.py:
```diff
@@ -187,11 +186,18 @@ def _run(self) -> str:
-    def _run(self, target_frame="map", source_frame="body_link") -> dict:
+    args_schema: Type[GetTransformInput] = GetTransformInput
+
+    def _run(self, target_frame="odom", source_frame="body_link") -> dict:
```
fixed in: https://github.com/RobotecAI/rai/pull/335/commits/13b4fe7ce392adafcfa8bb200e65c303ec83c348

boczekbartek replied:
It's set to string for the LLM. See: https://github.com/RobotecAI/rai/blob/13b4fe7ce392adafcfa8bb200e65c303ec83c348/src/rai/rai/tools/ros/native.py#L222

boczekbartek replied:
my last comment was to a different thread

boczekbartek committed:
fix(`NodeDiscovery`): repeated argument setting

boczekbartek commented on src/rai/rai/utils/ros.py:
```diff
@@ -12,19 +12,60 @@
+        self.topics_and_types: Dict[str, str] = dict()
+        self.services_and_types: Dict[str, str] = dict()
+        self.actions_and_types: Dict[str, str] = dict()
+        self.allowlist: Optional[List[str]] = allowlist
```
https://github.com/RobotecAI/rai/pull/335/commits/ad319b68858d8851b8ac34764a450f1373682be8

boczekbartek commented on src/rai/rai/node.py:
```diff
@@ -321,11 +334,123 @@ def get_msg_type(self, topic: str, n_tries: int = 5) -> Any:
+                    error = f"No message received in {timeout_sec} seconds from topic {topic}"
+                    self.get_logger().error(error)
+                    return error
+            finally:
```
It's set to string for the LLM. See: https://github.com/RobotecAI/rai/blob/13b4fe7ce392adafcfa8bb200e65c303ec83c348/src/rai/rai/tools/ros/native.py#L222

boczekbartek commented on src/rai/rai/node.py:
```diff
@@ -197,30 +206,54 @@ def _run_action(self, action_name, action_type, action_goal_args):
+            name="", bases=""
+        )  # arguments are not used
+        code_to_name = {v: k for k, v in name_to_code.items()}
+        return code_to_name.get(code, "UNKNOWN")
```
Actually the previous one worked well, I'll make a commit. I forgot about checking the instance

boczekbartek replied:
https://github.com/RobotecAI/rai/pull/335/commits/2aae9c881186ed6254f3c68f389b367b116f858a

boczekbartek committed:
refactor: ros2 action error code parsing

boczekbartek committed:
fix deprecated way of spinning the node in open-set vision

boczekbartek committed:
get transform synchronously

boczekbartek commented:
@maciejmajek I have noticed that open-set detection and `GetTransform` tools were not refactored properly and caused executor problems. Could you test again on your side?

boczekbartek committed:
rollback to fixed executor

boczekbartek committed:
add missing executor file

boczekbartek committed:
fix(`text_hmi`): `TaskAction` attribute is named `report`

boczekbartek committed:
chore: add missing licence header

boczekbartek committed:
avoid spinning the node in open-set vision tools

boczekbartek committed:
chore: pre-commit

boczekbartek committed:
test(`transport`): spin RaiBaseNode in test

boczekbartek commented:
@maciejmajek I rebased this PR and fixed failing transport test (in the implementation from this PR, the `RaiBaseNode` has to be spun for the node subscribers to work).

boczekbartek committed:
don't shutdown rclpy in rai node

boczekbartek committed:
rename Ros2*Helper to Ros2*Api and move to a separate file

boczekbartek committed:
add copyright notice
